libnvme/src/firmware.rs

```diff
@@ -0,0 +1,306 @@
+use std::{ffi::CString, mem::MaybeUninit};
+
+use libnvme_sys::nvme::*;
+use thiserror::Error;
+
+use crate::{
+    controller::{Controller, WriteLockedController},
+    controller_info::ControllerInfoIdentify,
+    error::LibraryError,
+    logpage::{LogPageInfo, LogPageName},
+    NvmeError,
+};
+
+#[derive(Debug, Error)]
+pub enum NvmeSlotError {
+    #[error("NVMe slots must be between 1 and 7 but got {0}")]
+    InvalidSlotNumber(usize),
+    #[error("NVMe device does not have slot {0}")]
+    DoesNotExisit(u32),
+    #[error("libnvme error: {0}")]
+    NvmeError(#[from] NvmeError),
+}
+
+// NOTE: we represent this is a u32 because `nvme_fw_commit_req_set_slot` takes
+// a uint32_t as the spec may change in the future to add additional slots.
+/// NVMe Slot Number.
+pub struct NvmeSlot(u32);
+
+// Rust's default interger type is `i32`.
+impl TryFrom<i32> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value as u32)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+// Setting a slot via `nvme_fw_commit_req_set_slot` uses a `u32`.
+impl TryFrom<u32> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+// The NVMe spec currently uses a u8 for active and next active slots.
+impl TryFrom<u8> for NvmeSlot {
+    type Error = NvmeSlotError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            1..=7 => Ok(NvmeSlot(value as _)),
+            invalid => Err(NvmeSlotError::InvalidSlotNumber(invalid as _)),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// The Firmwware log page for an NVMe controller.
+pub struct FirmwareLogPage {
+    /// The currently active firmware slot.
+    pub active_slot: u8,
+    /// The next slot that will be active on a controller reset.
+    pub next_active_slot: Option<u8>,
+    /// Slot 1 is read-only.
+    pub slot1_is_read_only: bool,
+    /// The number of firmware slots the controller has.
+    pub number_of_slots: usize,
+    firmware_slots: Vec<Option<String>>,
```

Comment:
Small nit - could we name this what the strings are? `firmware_slot_versions` or something like that

Reply:
Fixed in dff4996b